flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/ChangelogStorageMetricsTest.java

```diff
@@ -295,6 +349,39 @@ public void close() {
         }
     }
 
+    private static class FixedLatencyUploader implements StateChangeUploader {
+        private final long latency;
+
+        public FixedLatencyUploader(long latency) {
+            this.latency = latency;
+        }
+
+        @Override
+        public UploadTasksResult upload(Collection<UploadTask> tasks) throws IOException {
+            Map<UploadTask, Map<StateChangeSet, Long>> map = new HashMap<>();
+
+            try {
+                TimeUnit.MILLISECONDS.sleep(latency);
```

Comment:
Could you explain why `sleep` is needed here and why an existing `MaxAttemptUploader` can not be used?

Reply:
I want to simulate timeout situation instead of IOException. The difference between the two is that when MaxAttemptUploader is used, attemptNumber must be equal to totalAttempt, while when FixedLatencyUploader is used, attemptNumber is less than totalAttempt. First attempt will success with FixedLatencyUploader.

In fact, I considered merging testAttemptsPerUpload and testTotalAttemptsPerUpload together. This can be  implemented by adding the following to testTotalAttemptsPerUpload:

`
        HistogramStatistics attemptsPerUploadHistogram =
                metrics.getAttemptsPerUpload().getStatistics();
        assertThat(attemptsPerUploadHistogram.getMin()).isEqualTo(1);
        assertThat(attemptsPerUploadHistogram.getMax()).isEqualTo(1);
`

How do you think about @rkhachatryan ?

Reply:
Thanks for the explanation. 
In other words, `testAttemptsPerUpload` completes the last attempt and `testTotalAttemptsPerUpload` completes the first attempt, right?

IMO it's better to keep these tests separate, because the scenarios and the assertions are different.

As for the implementation of `testTotalAttemptsPerUpload`, how about waiting in the 1st attempt for the others to fail?
Something like `CountDownLatch` or `CompletableFuture` + `AtomicInteger` ?

Reply:
Thanks for your nice suggestion @rkhachatryan .

First, yes, testTotalAttemptsPerUpload completes the first attempt.

Second, I agree with keep these tests separate.

Third, the implementation of testTotalAttemptsPerUpload, I think you are right.
And I think we can implement waiting in 1st attempt for the others in MaxAttemptUploader. What do you think about ?

Reply:
I mean `MaxAttemptUploader` only have one behavior, just like you said, make first attempt waiting until all others attempt triggered.
Then both testAttemptsPerUpload and testTotalAttemptsPerUpload can use `MaxAttemptUploader`. WDYT ?

Reply:
In that case, `testAttemptsPerUpload` would assert that `metrics.AttemptsPerUpload == 1`, right?
That doesn't seem very useful because 1 is likely the number of attempts in any case, even without retries, so it wouldn't catch many bugs.

Reply:
Got it. I have changed the implementation of testTotalAttemptsPerUpload, and updated the pr.

Reply:
Thanks!